packages/headless/contributors/adding-a-sub-package.md

```diff
@@ -0,0 +1,36 @@
+# Adding a sub-package
+
+Headless provides exports through multiple sub packages. A sub-package groups together exports (i.e. controllers, actions, reducers, engines) that work together as a cohesive unit. By separating exports into sub-packages, it becomes clear to users of headless what exports are available to build a use-case.
+
+
+## To add a new sub-package:
+
+1. Create an entry file for your sub-package inside the `src` directory (e.g. case-assist.ts).
+2. Configure nodejs and browser bundles inside `rollup.config.js` for your entry file.
+3. Inside the `pkg` directory, create a new directory having the name of your sub-package.
+4. Inside the new directory, add a `package.json` file and fill in the paths to your bundled files and type definitions.
+
+```json
+pkg/case-assist/package.json
+
+{
+  "name": "case-assist",
+  "version": "1.0.0",
+  "description": "Headless Case Assist Module",
+  "main": "../../dist/case-assist/headless.js",
+  "module": "../../dist/case-assist/headless.esm.js",
+  "browser": "../../dist/browser/case-assist/headless.esm.js",
+  "types": "../../dist/definitions/case-assist.d.ts",
+  "license": "Apache-2.0"
+}
+```
+
+## Testing your sub-package:
+
+1. Build the headless project: `npm run build`.
+2. Create a tarball: `npm pack`.
+3. Install the tarball as a dependency of a different project: `npm i <path to the tarball>`.
+4. Import an export from your sub-package: `import {...} from '@coveo/headless/pkg/<sub-package>'`
```

Comment:
This would be possible, but with one possible draw-back: the `case-assist` directory holding the package.json file needs to be at the root level.

My concern was if headless has `n` packages, then the root level will become cluttered:
```
dist/
src/
package1/
package2/
package3/
...
```
By grouping the sub-packages under a `pkg` dir, the root is more tidy. Personally, I prefer the shorter path same as you, but I did this as a compromise.

I will check if it is possible to specify a base-path when creating the tarball, but assuming it is not, what do you think between adding the sub-packages to the root vs. grouping them under a dir?

Reply:
Updates:

Node 12.7 added support for an `exports` field to allow mapping one path to another.
- https://stackoverflow.com/questions/51313753/npm-package-json-base-root-property?noredirect=1#comment89604253_51313753
- https://github.com/jkrems/proposal-pkg-exports/#example

However, this does not yet work inside typescript projects. Support is expected as part of TS 4.3 due next month, but still not working as part of the beta.
- https://github.com/microsoft/TypeScript/issues/33079#issuecomment-817175406
- https://github.com/microsoft/TypeScript/issues/42762

Maybe we could start with top-level directories, and refactor once support is available. This would provide our ideal api, and avoid a breaking change.

I will look further tonight and make the decision tomorrow,